### PR TITLE
dts: bindings: interrupt-controller: GIC: Allow specifying version in DT

### PR DESCRIPTION
**Pull request by p-woj**
This PR separates the GIC binding into 3 version-specific ones and moves the GIC version from the SoCs' Kconfigs into their device trees. This brings the GIC driver more in line with other existing drivers and makes the device tree a more complete description of the hardware, which is useful for efforts relying on structured data such as [Renodepedia](https://zephyr-dashboard.renode.io/renodepedia/).

### CODE REVIEW

**Comment by povergoing** [comment]
LGTM, but someone who has downstream board patches have to update their DTS :)

**Review by stephanosio — COMMENTED** [review]
This looks reasonable; but, these changes need to be described in the release notes so that any downstream users using out-of-tree boards can handle the migration.

File: drivers/interrupt_controller/Kconfig.gic [hunk 1]
```diff
@@ -9,21 +9,21 @@ config GIC
 	bool
 
 config GIC_V1
-	bool
+	def_bool DT_HAS_ARM_GIC_V1_ENABLED
```
Comment (carlocaione) [inline drivers/interrupt_controller/Kconfig.gic:12]: For consistency this should be:
```
default y
depends on ...
```

**Comment by p-woj** [comment]
Thanks. I've added an explanation to the v3.4 release notes draft and converted the Kconfig def_bool to a dependency.

I now see an issue here though, and I'm going to need some advice on how to proceeed. Previously with the `def_bool` approach we (and indeed out-of-tree boards as well) could still use the old method of specifying only `arm,gic` as a compatible string for the GIC and setting its architecture version in Kconfig.

With the `default y` + dependency approach this is not the case anymore - if the GIC version is not present in the DT and the SoC's Kconfig selects it, the build will fail.

This is important not only for out-of-tree boards but also for Zephyr itself because it means that, in order to ensure a buildable state after each individual commit, we would need to first add the new compats to our DTs, then introduce the `DT_HAS` dependency, and only then remove the `select GIC_Vx`s from our Kconfigs. In that case, it might be beneficial to squash some of these commits together, as there are already quite a few considering the simplicity of the change, and this would almost double their number.

Looking forward to your suggestions.

**Comment by carlocaione** [comment]
> I now see an issue here though, and I'm going to need some advice on how to proceeed. Previously with the `def_bool` approach we (and indeed out-of-tree boards as well) could still use the old method of specifying only `arm,gic` as a compatible string for the GIC and setting its architecture version in Kconfig.

> With the default y + dependency approach this is not the case anymore - if the GIC version is not present in the DT and the SoC's Kconfig selects it, the build will fail.

This is interesting indeed. @mbolivar @galak do we have a process in place to deprecate a DT compatible?

**Review by lorc — DISMISSED** [review]
Finally someone did this! Thanks, it is much better in this way.

LGTM for Xen-related changes.

**Review by povergoing — DISMISSED** [review]

**Review by carlocaione — CHANGES_REQUESTED** [review]
LGTM but this is blocked on https://github.com/zephyrproject-rtos/zephyr/pull/58035#issuecomment-1554712792 for now

**Comment by mbolivar-nordic** [comment]
> do we have a process in place to deprecate a DT compatible?

No formal process, no. From https://github.com/zephyrproject-rtos/zephyr/pull/58035#issuecomment-1554712792, it's just a question of whether to squash some commits together to preserve bisectability, right? That seems reasonable to me.

**Comment by carlocaione** [comment]
> No formal process, no. From [#58035 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/58035#issuecomment-1554712792), it's just a question of whether to squash some commits together to preserve bisectability, right? That seems reasonable to me.

No, it is also a matter of breaking out-of-tree ARM64 boards, not just a bisectability issue.

**Comment by p-woj** [comment]
I've adjusted the Kconfig changes so that they don't break out-of-tree boards anymore. Now both ways of specifying the architecture version (`select GIC_V2` in Kconfig and `arm,gic-v2` in the DT) will work.

Also, having the version explicitly selected in Kconfig without the versioned compat in the DT prints a deprecation warning.

Please take a look @carlocaione

File: doc/releases/release-notes-3.4.rst [hunk 2]
```diff
@@ -179,6 +179,12 @@ Deprecated in this release
   functionality, enable
   :kconfig:option:`CONFIG_NRF_STORE_REBOOT_TYPE_GPREGRET`.
 
+* Setting the GIC architecture version by selecting
+  :kconfig:option:`CONFIG_GIC_V1`, :kconfig:option:`CONFIG_GIC_V2` and
+  :kconfig:option:`CONFIG_GIC_V3` directly in Kconfig has been deprecated.
+  The GIC version should now be specified by adding the appropriate compatible, for
+  example :dtcompatible:`arm,gic-v2`, to the GIC node in the device tree.
```
Comment (carlocaione) [inline doc/releases/release-notes-3.4.rst:186]: This should go now in the 3.5 release notes.

**Review by carlocaione — COMMENTED** [review]
LGTM. Just a note about the changelog

File: doc/releases/release-notes-3.4.rst (hunk 2 above)
Comment (p-woj) [reply doc/releases/release-notes-3.4.rst]: Should be good now

**Review by carlocaione — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Pull request merged**